src/core/cm/imagemanager/imagemanager.cpp

```diff
@@ -0,0 +1,949 @@
+{
+    for (const auto& itemImage : itemImages) {
+        if (auto err = status.mStatuses.EmplaceBack(); !err.IsNone()) {
+            LOG_ERR() << "Can't add update item status" << Log::Field("ID", status.mID)
+                      << Log::Field("Version", status.mVersion) << Log::Field("Error", err);
+            return err;
+        }
+
+        auto& itemStatus = status.mStatuses.Back();
+
+        itemStatus.mImageID = itemImage.mImageID;
+        itemStatus.mState   = error.IsNone() ? state.GetValue() : ImageStateEnum::eFailed;
+        itemStatus.mError   = error;
+
+        NotifyImageStatusChangedListeners(itemStatus);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ImageManager::InstallUpdateItem(const UpdateItemInfo& itemInfo,
+    const Array<cloudprotocol::CertificateInfo>&            certificates,
+    const Array<cloudprotocol::CertificateChainInfo>& certificateChains, UpdateItemStatus& status)
+{
+    LOG_DBG() << "Install update item" << Log::Field("ID", itemInfo.mID) << Log::Field("Version", itemInfo.mVersion)
+              << Log::Field("Images", itemInfo.mImages.Size());
+
+    Error err;
+    auto  items = MakeUnique<StaticArray<storage::ItemInfo, storage::cMaxItemVersions>>(&mAllocator);
+
+    auto fillStatus = DeferRelease(&err, [&](Error* err) {
+        status.mID      = itemInfo.mID;
+        status.mVersion = itemInfo.mVersion;
+
+        if (err->IsNone() || !status.mStatuses.IsEmpty()) {
+            return;
+        }
+
+        for (const auto& image : itemInfo.mImages) {
+            if (auto errEmplace = status.mStatuses.EmplaceBack(); !errEmplace.IsNone()) {
+                LOG_ERR() << "Can't emplace back update item status" << Log::Field("ID", itemInfo.mID)
+                          << Log::Field("Version", itemInfo.mVersion) << Log::Field("Error", errEmplace);
+
+                return;
+            }
+
+            auto& itemStatus = status.mStatuses.Back();
+
+            itemStatus.mImageID = image.mImage.mImageID;
+            itemStatus.mState   = ImageStateEnum::eFailed;
+            itemStatus.mError   = *err;
+
+            NotifyImageStatusChangedListeners(itemStatus);
+        }
+    });
+
+    if (err = mStorage->GetItemVersionsByID(itemInfo.mID, *items); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (err = ValidateActiveVersionItem(itemInfo, *items); !err.IsNone()) {
+        if (!err.Is(ErrorEnum::eNotFound)) {
+            return err;
+        }
+
+        if (err = ValidateCachedVersionItem(itemInfo, *items); !err.IsNone()) {
+            if (!err.Is(ErrorEnum::eNotFound)) {
+                return err;
+            }
+        }
+    }
+
+    size_t totalSize {};
+    for (const auto& image : itemInfo.mImages) {
+        totalSize += image.mSize;
+    }
+
+    UniquePtr<spaceallocator::SpaceItf> space;
+
+    Tie(space, err) = mSpaceAllocator->AllocateSpace(totalSize);
```

Comment:
space allocator should be used before decryption. Otherwise it doesn't make sense.